File: app/services/sitemaps.py
from __future__ import annotations

from datetime import datetime, timezone

import httpx
from sqlalchemy import select
from sqlalchemy.orm import Session

from app.google import search_console as gsc
from app.google.errors import GoogleApiError
from app.google.oauth import get_access_token
from app.models import IndexJob, JobStatus, JobType, Site, Sitemap, User
from app.services import sitemap_parser
from app.services.activity import log_event
from app.services.urls import add_urls


def _now() -> datetime:
    return datetime.now(timezone.utc).replace(tzinfo=None)


def sync_from_gsc(db: Session, user: User, site: Site) -> dict:
    """Mirror the sitemap list Google already knows about."""
    token = get_access_token(db, user)
    entries = gsc.list_sitemaps(token, site.property_url)

    created = 0
    updated = 0
    for entry in entries:
        path = entry.get("path")
        if not path:
            continue
        sitemap = db.scalar(
            select(Sitemap).where(Sitemap.site_id == site.id, Sitemap.path == path)
        )
        if sitemap is None:
            sitemap = Sitemap(site_id=site.id, path=path, source="gsc")
            db.add(sitemap)
            created += 1
        else:
            updated += 1

        sitemap.is_pending = bool(entry.get("isPending"))
        sitemap.is_sitemaps_index = bool(entry.get("isSitemapsIndex"))
        sitemap.warnings = int(entry.get("warnings", 0) or 0)
        sitemap.errors = int(entry.get("errors", 0) or 0)
        contents = entry.get("contents") or []
        sitemap.url_count = sum(int(c.get("submitted", 0) or 0) for c in contents)
        sitemap.last_submitted_at = _parse_ts(entry.get("lastSubmitted"))
        sitemap.last_downloaded_at = _parse_ts(entry.get("lastDownloaded"))

    db.commit()
    return {"created": created, "updated": updated, "total": len(entries)}


def discover_sitemaps(db: Session, site: Site) -> list[str]:
    """Probe well-known sitemap paths and robots.txt."""
    found: list[str] = []
    for candidate in sitemap_parser.guess_sitemap_urls(site.home_url):
        try:
            with httpx.Client(timeout=12, follow_redirects=True) as client:
                response = client.head(
                    candidate, headers={"User-Agent": sitemap_parser.USER_AGENT}
                )
                if response.status_code >= 400:
                    response = client.get(
                        candidate, headers={"User-Agent": sitemap_parser.USER_AGENT}
                    )
            if response.status_code < 400:
                found.append(candidate)
        except httpx.HTTPError:
            continue

    for path in found:
        existing = db.scalar(
            select(Sitemap).where(Sitemap.site_id == site.id, Sitemap.path == path)
        )
        if existing is None:
            db.add(Sitemap(site_id=site.id, path=path, source="discovered"))
    db.commit()
    return found


def add_sitemap(db: Session, site: Site, path: str) -> Sitemap:
    clean = path.strip()
    if not clean.startswith("http"):
        clean = site.home_url.rstrip("/") + "/" + clean.lstrip("/")
    existing = db.scalar(select(Sitemap).where(Sitemap.site_id == site.id, Sitemap.path == clean))
    if existing:
        return existing
    sitemap = Sitemap(site_id=site.id, path=clean, source="manual")
    db.add(sitemap)
    db.commit()
    db.refresh(sitemap)
    return sitemap


def submit_to_google(db: Session, user: User, site: Site, sitemap: Sitemap) -> IndexJob:
    job = IndexJob(
        site_id=site.id,
        target=sitemap.path,
        job_type=JobType.SITEMAP_SUBMIT,
        status=JobStatus.RUNNING,
    )
    db.add(job)
    db.flush()

    try:
        token = get_access_token(db, user)
        gsc.submit_sitemap(token, site.property_url, sitemap.path)
        sitemap.last_submitted_at = _now()
        job.status = JobStatus.SUCCESS
        job.message = "Sitemapa zgloszona do Google Search Console"
    except GoogleApiError as exc:
        job.status = JobStatus.FAILED
        job.message = str(exc)
    finally:
        job.finished_at = _now()
        db.commit()
    return job


def delete_from_google(db: Session, user: User, site: Site, sitemap: Sitemap) -> IndexJob:
    job = IndexJob(
        site_id=site.id,
        target=sitemap.path,
        job_type=JobType.SITEMAP_DELETE,
        status=JobStatus.RUNNING,
    )
    db.add(job)
    db.flush()

    try:
        token = get_access_token(db, user)
        gsc.delete_sitemap(token, site.property_url, sitemap.path)
        job.status = JobStatus.SUCCESS
        job.message = "Sitemapa usunieta z Google Search Console"
    except GoogleApiError as exc:
        job.status = JobStatus.FAILED
        job.message = str(exc)
    finally:
        job.finished_at = _now()
        db.commit()
    return job


def scan_sitemap(db: Session, site: Site, sitemap: Sitemap) -> dict:
    """Download the sitemap and import every URL it contains."""
    result = sitemap_parser.crawl_sitemap(sitemap.path)
    if result.error and not result.entries:
        sitemap.errors = (sitemap.errors or 0) + 1
        db.commit()
        return {"added": 0, "duplicates": 0, "error": result.error, "found": 0}

    valid = [
        entry
        for entry in result.entries
        if sitemap_parser.url_belongs_to_site(entry.url, site.home_url, site.is_domain_property)
    ]
    lastmod_map = {entry.url: entry.lastmod for entry in valid if entry.lastmod}
    outcome = add_urls(
        db, site, [entry.url for entry in valid], source="sitemap", lastmod_map=lastmod_map
    )

    sitemap.url_count = len(valid)
    sitemap.last_downloaded_at = _now()
    sitemap.is_sitemaps_index = result.is_index
    site.last_scan_at = _now()
    db.commit()

    outcome["found"] = len(result.entries)
    outcome["error"] = result.error
    return outcome


def scan_all_sitemaps(db: Session, site: Site) -> dict:
    totals = {"added": 0, "duplicates": 0, "found": 0, "sitemaps": 0, "errors": []}
    sitemaps = list(
        db.scalars(
            select(Sitemap).where(Sitemap.site_id == site.id, Sitemap.auto_sync.is_(True))
        )
    )
    if not sitemaps:
        discover_sitemaps(db, site)
        sitemaps = list(db.scalars(select(Sitemap).where(Sitemap.site_id == site.id)))

    for sitemap in sitemaps:
        outcome = scan_sitemap(db, site, sitemap)
        totals["added"] += outcome["added"]
        totals["duplicates"] += outcome["duplicates"]
        totals["found"] += outcome.get("found", 0)
        totals["sitemaps"] += 1
        if outcome.get("error"):
            totals["errors"].append(f"{sitemap.path}: {outcome['error']}")

    log_event(
        db,
        f"Skan sitemap dla {site.display_name}: +{totals['added']} nowych URL-i.",
        workspace_id=site.workspace_id,
        category="sitemap",
        details=totals,
        commit=True,
    )
    return totals


def _parse_ts(value: str | None) -> datetime | None:
    if not value:
        return None
    try:
        return datetime.fromisoformat(value.replace("Z", "+00:00")).replace(tzinfo=None)
    except ValueError:
        return None

File: src/services/sitemaps.ts
import { db } from "../db/index.js";
import { JobStatus, JobType, isDomainProperty, type Site, type Sitemap } from "../db/types.js";
import { GoogleApiError } from "../google/errors.js";
import { getAccessToken } from "../google/oauth.js";
import * as gsc from "../google/searchConsole.js";
import { sha256 } from "../lib/crypto.js";
import { parseDate } from "../lib/dates.js";
import { logEvent } from "./activity.js";
import * as sitemapParser from "./sitemapParser.js";
import { addUrls } from "./urls.js";

const USER_AGENT = "IndexMeNow/1.0 (+sitemap-crawler)";

function toInt(value: unknown): number {
  const parsed = Number.parseInt(String(value ?? 0), 10);
  return Number.isFinite(parsed) ? parsed : 0;
}

export function listSitemaps(siteId: number): Promise<Sitemap[]> {
  return db.selectFrom("sitemaps").selectAll().where("site_id", "=", siteId).orderBy("id").execute();
}

export function getSitemap(sitemapId: number, siteId: number): Promise<Sitemap | undefined> {
  return db
    .selectFrom("sitemaps")
    .selectAll()
    .where("id", "=", sitemapId)
    .where("site_id", "=", siteId)
    .executeTakeFirst();
}

/** Odzwierciedla liste sitemap, ktore Google juz zna. */
export async function syncFromGsc(userId: number, site: Site) {
  const token = await getAccessToken(userId);
  const entries = await gsc.listSitemaps(token, site.property_url);

  let created = 0;
  let updated = 0;

  for (const entry of entries) {
    const path = entry.path;
    if (!path) continue;
    const hash = sha256(path);

    const values = {
      is_pending: Boolean(entry.isPending),
      is_sitemaps_index: Boolean(entry.isSitemapsIndex),
      warnings: toInt(entry.warnings),
      errors: toInt(entry.errors),
      url_count: (entry.contents ?? []).reduce((sum, c) => sum + toInt(c.submitted), 0),
      last_submitted_at: parseDate(entry.lastSubmitted),
      last_downloaded_at: parseDate(entry.lastDownloaded),
    };

    const existing = await db
      .selectFrom("sitemaps")
      .select("id")
      .where("site_id", "=", site.id)
      .where("path_hash", "=", hash)
      .executeTakeFirst();

    if (existing) {
      await db.updateTable("sitemaps").set(values).where("id", "=", existing.id).execute();
      updated += 1;
    } else {
      await db
        .insertInto("sitemaps")
        .values({ site_id: site.id, path, path_hash: hash, source: "gsc", ...values })
        .execute();
      created += 1;
    }
  }

  return { created, updated, total: entries.length };
}

/** Sprawdza typowe sciezki sitemap i to, co ogłasza robots.txt. */
export async function discoverSitemaps(site: Site): Promise<string[]> {
  const found: string[] = [];

  for (const candidate of await sitemapParser.guessSitemapUrls(site.home_url)) {
    try {
      // Czesc serwerow nie obsluguje HEAD - wtedy sprawdzamy przez GET.
      let response = await fetch(candidate, {
        method: "HEAD",
        headers: { "User-Agent": USER_AGENT },
        redirect: "follow",
        signal: AbortSignal.timeout(12_000),
      });
      if (!response.ok) {
        response = await fetch(candidate, {
          headers: { "User-Agent": USER_AGENT },
          redirect: "follow",
          signal: AbortSignal.timeout(12_000),
        });
      }
      if (response.ok) found.push(candidate);
    } catch {
      continue;
    }
  }

  for (const path of found) {
    const hash = sha256(path);
    const existing = await db
      .selectFrom("sitemaps")
      .select("id")
      .where("site_id", "=", site.id)
      .where("path_hash", "=", hash)
      .executeTakeFirst();
    if (!existing) {
      await db
        .insertInto("sitemaps")
        .values({ site_id: site.id, path, path_hash: hash, source: "discovered" })
        .execute();
    }
  }

  return found;
}

export async function addSitemap(site: Site, path: string): Promise<Sitemap> {
  let clean = path.trim();
  if (!clean.startsWith("http")) {
    clean = `${site.home_url.replace(/\/+$/, "")}/${clean.replace(/^\/+/, "")}`;
  }
  const hash = sha256(clean);

  const existing = await db
    .selectFrom("sitemaps")
    .selectAll()
    .where("site_id", "=", site.id)
    .where("path_hash", "=", hash)
    .executeTakeFirst();
  if (existing) return existing;

  const inserted = await db
    .insertInto("sitemaps")
    .values({ site_id: site.id, path: clean, path_hash: hash, source: "manual" })
    .executeTakeFirst();

  return db
    .selectFrom("sitemaps")
    .selectAll()
    .where("id", "=", Number(inserted.insertId))
    .executeTakeFirstOrThrow();
}

export async function deleteSitemap(sitemapId: number, siteId: number): Promise<void> {
  await db.deleteFrom("sitemaps").where("id", "=", sitemapId).where("site_id", "=", siteId).execute();
}

async function runSitemapJob(
  site: Site,
  sitemap: Sitemap,
  jobType: JobType,
  action: (token: string) => Promise<unknown>,
  successMessage: string,
  userId: number,
) {
  const inserted = await db
    .insertInto("index_jobs")
    .values({
      site_id: site.id,
      target: sitemap.path.slice(0, 2048),
      job_type: jobType,
      status: JobStatus.RUNNING,
    })
    .executeTakeFirst();
  const jobId = Number(inserted.insertId);

  let status: JobStatus = JobStatus.SUCCESS;
  let message = successMessage;
  try {
    const token = await getAccessToken(userId);
    await action(token);
    if (jobType === JobType.SITEMAP_SUBMIT) {
      await db
        .updateTable("sitemaps")
        .set({ last_submitted_at: new Date() })
        .where("id", "=", sitemap.id)
        .execute();
    }
  } catch (error) {
    status = JobStatus.FAILED;
    message = error instanceof GoogleApiError ? error.toString() : String(error);
  }

  await db
    .updateTable("index_jobs")
    .set({ status, message, finished_at: new Date() })
    .where("id", "=", jobId)
    .execute();

  return { status, message };
}

export function submitToGoogle(userId: number, site: Site, sitemap: Sitemap) {
  return runSitemapJob(
    site,
    sitemap,
    JobType.SITEMAP_SUBMIT,
    (token) => gsc.submitSitemap(token, site.property_url, sitemap.path),
    "Sitemapa zgloszona do Google Search Console",
    userId,
  );
}

export function deleteFromGoogle(userId: number, site: Site, sitemap: Sitemap) {
  return runSitemapJob(
    site,
    sitemap,
    JobType.SITEMAP_DELETE,
    (token) => gsc.deleteSitemap(token, site.property_url, sitemap.path),
    "Sitemapa usunieta z Google Search Console",
    userId,
  );
}

export interface ScanOutcome {
  added: number;
  duplicates: number;
  refreshed: number;
  found: number;
  error?: string | null;
}

/** Pobiera sitemape i importuje wszystkie zawarte w niej adresy. */
export async function scanSitemap(site: Site, sitemap: Sitemap): Promise<ScanOutcome> {
  const result = await sitemapParser.crawlSitemap(sitemap.path);

  if (result.error && result.entries.length === 0) {
    await db
      .updateTable("sitemaps")
      .set({ errors: (sitemap.errors ?? 0) + 1 })
      .where("id", "=", sitemap.id)
      .execute();
    return { added: 0, duplicates: 0, refreshed: 0, found: 0, error: result.error };
  }

  // Sitemapa moze wymieniac adresy z innych domen - Google odrzuci takie
  // zgloszenie, wiec odsiewamy je zanim trafia do bazy.
  const domainProperty = isDomainProperty(site);
  const valid = result.entries.filter((entry) =>
    sitemapParser.urlBelongsToSite(entry.url, site.home_url, domainProperty),
  );

  const lastmodMap = new Map<string, Date>();
  for (const entry of valid) {
    if (entry.lastmod) lastmodMap.set(entry.url, entry.lastmod);
  }

  const outcome = await addUrls(
    site,
    valid.map((entry) => entry.url),
    "sitemap",
    lastmodMap,
  );

  const now = new Date();
  await db
    .updateTable("sitemaps")
    .set({
      url_count: valid.length,
      last_downloaded_at: now,
      is_sitemaps_index: result.isIndex,
    })
    .where("id", "=", sitemap.id)
    .execute();
  await db.updateTable("sites").set({ last_scan_at: now }).where("id", "=", site.id).execute();

  return { ...outcome, found: result.entries.length, error: result.error ?? null };
}

export interface ScanTotals {
  added: number;
  duplicates: number;
  found: number;
  sitemaps: number;
  errors: string[];
}

export async function scanAllSitemaps(site: Site): Promise<ScanTotals> {
  const totals: ScanTotals = { added: 0, duplicates: 0, found: 0, sitemaps: 0, errors: [] };

  let sitemaps = await db
    .selectFrom("sitemaps")
    .selectAll()
    .where("site_id", "=", site.id)
    .where("auto_sync", "=", true)
    .execute();

  if (sitemaps.length === 0) {
    await discoverSitemaps(site);
    sitemaps = await listSitemaps(site.id);
  }

  for (const sitemap of sitemaps) {
    const outcome = await scanSitemap(site, sitemap);
    totals.added += outcome.added;
    totals.duplicates += outcome.duplicates;
    totals.found += outcome.found;
    totals.sitemaps += 1;
    if (outcome.error) totals.errors.push(`${sitemap.path}: ${outcome.error}`);
  }

  await logEvent(`Skan sitemap dla ${site.display_name}: +${totals.added} nowych URL-i.`, {
    workspaceId: site.workspace_id,
    category: "sitemap",
    details: totals,
  });
  return totals;
}
